File: app/src/main/java/org/ole/planet/myplanet/utilities/DialogUtils.java
package org.ole.planet.myplanet.utilities;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.provider.Settings;
import android.view.View;

import androidx.appcompat.app.AlertDialog;

import com.google.android.material.snackbar.Snackbar;

import org.ole.planet.myplanet.MainApplication;
import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.datamanager.MyDownloadService;
import org.ole.planet.myplanet.datamanager.Service;
import org.ole.planet.myplanet.model.MyPlanet;

import java.util.ArrayList;


public class DialogUtils {

    public static ProgressDialog getProgressDialog(final Context context) {
        final ProgressDialog prgDialog = new ProgressDialog(context);
        prgDialog.setTitle(context.getString(R.string.downloading_file));
        prgDialog.setMax(100);
        prgDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
        prgDialog.setCancelable(false);
        prgDialog.setButton(DialogInterface.BUTTON_POSITIVE, context.getString(R.string.dismiss), (dialogInterface, i) -> prgDialog.dismiss());
        prgDialog.setButton(DialogInterface.BUTTON_NEGATIVE, context.getString(R.string.stop_download), (dialogInterface, i) -> context.stopService(new Intent(context, MyDownloadService.class)));
        return prgDialog;
    }

    public static void showError(ProgressDialog prgDialog, String message) {
        prgDialog.setTitle(message);
        if (prgDialog.getButton(ProgressDialog.BUTTON_NEGATIVE) != null)
            prgDialog.getButton(ProgressDialog.BUTTON_NEGATIVE).setEnabled(false);
    }

    public static void showWifiSettingDialog(final Context context) {
        if (!NetworkUtils.isWifiBluetoothEnabled())
            return;
        showDialog(context);
    }

    private static void showDialog(final Context context) {
        if (MainApplication.syncFailedCount > 3) {
            AlertDialog.Builder pd = new AlertDialog.Builder(context);
            String message = NetworkUtils.isBluetoothEnabled() ? "Bluetooth " : "";
            message += NetworkUtils.isWifiEnabled() ? "Wifi " : "";
            message += " is on please turn of to save battery";
            pd.setMessage(message);
            pd.setPositiveButton("Go to settings", (dialogInterface, i) -> {
                MainApplication.syncFailedCount = 0;
                Intent intent = new Intent(Settings.ACTION_WIFI_SETTINGS);
                context.startActivity(intent);
            }).setNegativeButton(context.getString(R.string.cancel), null);
            pd.setCancelable(false);
            AlertDialog d = pd.create();
//            d.getWindow().setType(WindowManager.LayoutParams.TYPE_SYSTEM_ALERT);
            d.show();
        }
    }

    public static void showSnack(View v, String s) {
        if (v != null)
            Snackbar.make(v, s, Snackbar.LENGTH_LONG).show();
    }

    public static void showAlert(Context context, String title, String message) {
        new AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(R.string.dismiss, null)
                .show();
    }

    public static AlertDialog getAlertDialog(Context context, String message,String pos, DialogInterface.OnClickListener listener) {
        return new AlertDialog.Builder(context)
                .setMessage(message)
                .setIcon(R.drawable.courses)
                .setPositiveButton(pos, listener).setNegativeButton("Cancel", null).show();
    }

    public static void showCloseAlert(Context context, String title, String message) {
        new AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(R.string.close, null)
                .show();
    }

    public static AlertDialog getAlertDialog(Context context, String title, View v) {
        return new AlertDialog.Builder(context)
                .setTitle(title)
                .setIcon(R.drawable.ic_edit)
                .setView(v)
                .setPositiveButton("Submit", null).setNegativeButton("Cancel", null).show();
    }

    public static AlertDialog.Builder getUpdateDialog(Context context, MyPlanet info, ProgressDialog progressDialog) {
        return new AlertDialog.Builder(context).setTitle("New version of my planet available")
                .setMessage("Download first to continue.")
                .setNeutralButton("Upgrade(Local)", (dialogInterface, i) -> {
                    startDownloadUpdate(context, Utilities.getApkUpdateUrl(info.getLocalapkpath()), progressDialog);
                })
                .setPositiveButton("Upgrade", (dialogInterface, i) -> {
                    startDownloadUpdate(context, info.getApkpath(), progressDialog);
                });

    }

    public static void startDownloadUpdate(Context context, String path, ProgressDialog progressDialog) {
        new Service(MainApplication.context).checkCheckSum(new Service.ChecksumCallback() {
            @Override
            public void onMatch() {
                Utilities.toast(MainApplication.context, "Apk already exists");
                FileUtils.installApk(context, path);
            }

            @Override
            public void onFail() {
                ArrayList url = new ArrayList();
                url.add(path);
                if (progressDialog != null) {
                    progressDialog.setMessage("Downloading file...");
                    progressDialog.setCancelable(false);
                    progressDialog.show();
                }
                Utilities.openDownloadService(context, url);
            }
        }, path);

    }
}

File: app/src/main/java/org/ole/planet/myplanet/utilities/DialogUtils.kt
package org.ole.planet.myplanet.utilities

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.provider.Settings
import android.view.ContextThemeWrapper
import android.view.LayoutInflater
import android.view.View
import androidx.appcompat.app.AlertDialog
import com.google.android.material.snackbar.Snackbar
import org.ole.planet.myplanet.MainApplication
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.DialogProgressBinding
import org.ole.planet.myplanet.datamanager.MyDownloadService
import org.ole.planet.myplanet.datamanager.Service
import org.ole.planet.myplanet.model.MyPlanet
import org.ole.planet.myplanet.service.UserProfileDbHandler
import org.ole.planet.myplanet.ui.userprofile.BecomeMemberActivity

object DialogUtils {
    @JvmStatic
    fun getProgressDialog(context: Context): CustomProgressDialog {
        val prgDialog = CustomProgressDialog(context)
        prgDialog.setTitle(context.getString(R.string.downloading_file))
        prgDialog.setMax(100)
        prgDialog.setNegativeButton(context.getString(R.string.stop_download), isVisible = true) {
            context.stopService(Intent(context, MyDownloadService::class.java))
            prgDialog.dismiss()
        }
        return prgDialog
    }

    fun guestDialog(context: Context) {
        val profileDbHandler = UserProfileDbHandler(context)
        val builder = android.app.AlertDialog.Builder(context, R.style.CustomAlertDialog)
        builder.setTitle(context.getString(R.string.become_a_member))
        builder.setMessage(context.getString(R.string.to_access_this_feature_become_a_member))
        builder.setCancelable(false)
        builder.setPositiveButton(context.getString(R.string.become_a_member), null)
        builder.setNegativeButton(context.getString(R.string.cancel), null)

        val dialog = builder.create()
        dialog.show()

        val becomeMember = dialog.getButton(android.app.AlertDialog.BUTTON_POSITIVE)
        val cancel = dialog.getButton(android.app.AlertDialog.BUTTON_NEGATIVE)
        becomeMember.contentDescription = context.getString(R.string.confirm_membership)
        cancel.contentDescription = context.getString(R.string.cancel)

        becomeMember.setOnClickListener {
            val guest = true
            val intent = Intent(context, BecomeMemberActivity::class.java)
            intent.putExtra("username", profileDbHandler.userModel?.name)
            intent.putExtra("guest", guest)
            context.startActivity(intent)
        }
        cancel.setOnClickListener {
            dialog.dismiss()
        }
    }

    @JvmStatic
    fun showError(prgDialog: CustomProgressDialog?, message: String?) {
        prgDialog?.setTitle(message)
        prgDialog?.disableNegativeButton()
    }

    @JvmStatic
    fun showWifiSettingDialog(context: Context) {
        if (!NetworkUtils.isWifiBluetoothEnabled()) return
        showDialog(context)
    }

    private fun showDialog(context: Context) {
        if (MainApplication.syncFailedCount > 3) {
            val pd = AlertDialog.Builder(context, R.style.AlertDialogTheme)
            var message = ""
            if (NetworkUtils.isBluetoothEnabled()) message += "Bluetooth"
            if (NetworkUtils.isWifiEnabled()) {
                if (message.isNotEmpty()) message += " and "
                    message += "Wi-Fi"
            }
            if (message.isNotEmpty()) {
            message += context.getString(R.string.is_on_please_turn_of_to_save_battery)
            } else {
            message = context.getString(R.string.is_on_please_turn_of_to_save_battery)
            }
            pd.setMessage(message)
            pd.setPositiveButton(context.getString(R.string.go_to_settings)) { _, _ ->
                MainApplication.syncFailedCount = 0
                val intent = Intent(Settings.ACTION_WIFI_SETTINGS)
                context.startActivity(intent)
            }.setNegativeButton(context.getString(R.string.cancel), null)
            pd.setCancelable(false)
            val d = pd.create()
            d.show()
        }
    }

    @JvmStatic
    fun showSnack(v: View?, s: String?) {
        if (v != null) {
            s?.let { Snackbar.make(v, it, Snackbar.LENGTH_LONG).show() }
        }
    }

    @JvmStatic
    fun showAlert(context: Context?, title: String?, message: String?) {
        if (context is Activity && !context.isFinishing) {
            AlertDialog.Builder(context, R.style.AlertDialogTheme)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(R.string.finish, null)
                .show()
        }
    }

    @JvmStatic
    fun getAlertDialog(context: Context, message: String, pos: String, listener: DialogInterface.OnClickListener?): AlertDialog {
        return AlertDialog.Builder(ContextThemeWrapper(context, R.style.CustomAlertDialog))
            .setMessage(message)
            .setIcon(R.drawable.courses)
            .setPositiveButton(pos, listener)
            .setNegativeButton(R.string.button_cancel, null)
            .show()
    }

    @JvmStatic
    fun showCloseAlert(context: Context, title: String?, message: String) {
        AlertDialog.Builder(context, R.style.CustomAlertDialog)
            .setTitle(title)
            .setMessage(message)
            .setPositiveButton(R.string.close, null)
            .show()
    }

    @JvmStatic
    fun getAlertDialog(context: Context, title: String, v: View): AlertDialog {
        return AlertDialog.Builder(ContextThemeWrapper(context, R.style.CustomAlertDialog))
            .setTitle(title)
            .setIcon(R.drawable.ic_edit)
            .setView(v)
            .setPositiveButton(R.string.submit, null)
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    @JvmStatic
    fun getUpdateDialog(context: Context, info: MyPlanet?, progressDialog: CustomProgressDialog?): AlertDialog.Builder {
        return AlertDialog.Builder(context, R.style.CustomAlertDialog)
            .setTitle(R.string.new_version_of_my_planet_available)
            .setMessage(R.string.download_first_to_continue)
            .setNeutralButton(R.string.upgrade_local) { _, _ ->
                startDownloadUpdate(context, Utilities.getApkUpdateUrl(info?.localapkpath), progressDialog)
            }
            .setPositiveButton(R.string.upgrade) { _, _ ->
                info?.apkpath?.let { startDownloadUpdate(context, it, progressDialog) }
            }
    }

    @JvmStatic
    fun startDownloadUpdate(context: Context, path: String, progressDialog: CustomProgressDialog?) {
        Service(MainApplication.context).checkCheckSum(object : Service.ChecksumCallback {
            override fun onMatch() {
                Utilities.toast(MainApplication.context, context.getString(R.string.apk_already_exists))
                FileUtils.installApk(context, path)
            }

            override fun onFail() {
                val urls = arrayListOf(path)
                if (progressDialog != null) {
                    progressDialog.setText(context.getString(R.string.downloading_file))
                    progressDialog.setCancelable(false)
                    progressDialog.show()
                }
                MyDownloadService.startService(context, "$urls", false)
            }
        }, path)
    }

    @JvmStatic
    fun getCustomProgressDialog(context: Context): CustomProgressDialog {
        return CustomProgressDialog(context)
    }

    class CustomProgressDialog(context: Context) {
        private val binding: DialogProgressBinding = DialogProgressBinding.inflate(LayoutInflater.from(context))
        private val dialogBuilder: AlertDialog.Builder = AlertDialog.Builder(context, R.style.CustomAlertDialog)
        private val progressBar = binding.progressBar
        private val progressText = binding.progressText
        private val progressTitle = binding.progressTitle
        private var dialog: AlertDialog? = null
        private var positiveButtonAction: (() -> Unit)? = null
        private var negativeButtonAction: (() -> Unit)? = null

        init {
            dialogBuilder.setView(binding.root)
            dialogBuilder.setCancelable(false)
            binding.buttonPositive.setOnClickListener {
                positiveButtonAction?.invoke()
            }
            binding.buttonNegative.setOnClickListener {
                negativeButtonAction?.invoke()
            }
        }
        fun setPositiveButton(text: String, isVisible: Boolean = true, listener: () -> Unit) {
            binding.buttonPositive.text = text
            positiveButtonAction = listener
            binding.buttonPositive.visibility = if (isVisible) View.VISIBLE else View.GONE
        }

        fun setNegativeButton(text: String = "Cancel", isVisible: Boolean = true, listener: () -> Unit) {
            binding.buttonNegative.text = text
            negativeButtonAction = listener
            binding.buttonNegative.visibility = if (isVisible) View.VISIBLE else View.GONE
        }

        fun show() {
            if (dialog == null) {
                dialog = dialogBuilder.create()
            }
            if (dialog?.isShowing == false) {
                dialog?.show()
            }
        }

        fun dismiss() {
            dialog?.dismiss()
        }

        fun setCancelable(state: Boolean) {
            dialog?.setCancelable(state)
        }

        private fun setIndeterminate() {
            progressBar.isIndeterminate = false
        }

        fun setMax(maxValue: Int) {
            progressBar.max = maxValue
        }

        fun setProgress(value: Int) {
            setIndeterminate()
            progressBar.progress = value
        }

        fun setText(text: String) {
            progressText.text = text
            progressText.visibility = View.VISIBLE
        }

        fun setTitle(text: String?) {
            progressTitle.visibility = View.VISIBLE
            progressTitle.text = text
        }
        fun isShowing(): Boolean {
            return dialog?.isShowing == true
        }

        fun disableNegativeButton() {
            binding.buttonNegative.isEnabled = false
        }
    }
}
